### ISSUE TITLE
Callbacks not being triggered when job is retried

### ISSUE THREAD

**Issue by djhoughton** [Shopify/job-iteration#33]
We have a `Job` that uses the `on_start` callback and has concurrency set to `1` so only one job can run per shop at any given time. 

But we're getting cases where the `on_start` is never being called even though the job is running. 

We believe this is what is happening:
- `Job1` starts
- `Job2` tries to start
- It can't start since one is already running so we wait for 5 seconds before retrying
- `Job2.executions` is incremented by 1
- `Job1` finishes
- `Job2` tries to run
- It is able to run since the first job is completed
- `executions` is now at 2 so the callbacks are not triggered [(code)](https://github.com/Shopify/job-iteration/blob/55b4640129a47d0c638cc7810741b333bf8b968c/lib/job-iteration/iteration.rb#L104)

Feel free to DM me for Splunk queries or more details about our job and setup.

**Comment by DrewMartin** [comment]
@Shopify/job-patterns

**Comment by kirs** [comment]
I'm guessing this is kind of a way how `concurrency` and `executions` works.

Is it an option for you to implement the callback based on records from the business logic?

E.g.

```ruby
class ImportJob < Base
  before_action do
    import = arguments[0]['import_id']
    if !import.started?
      # do some action
      import.update(started: true)
    end
  end
end
```

**Comment by djhoughton** [comment]
Yes, we've worked around it by always calling our callback in the `build_enumerator` method of the job and having that method return early if the work is already done. 👍 
https://github.com/Shopify/shopify/pull/197079

**Comment by DrewMartin** [comment]
It's easy enough in this case to remove the callback and replace it with inline code, the bigger issue is that this can happen to other jobs that use `Concurrency`. Can we at least raise if they try to use `on_start` callbacks when they use `Concurrency` so it doesn't crop up unexpectedly?

**Comment by kirs** [comment]
Agree this might be unexpected and confusing.

I'd suggest we remove `on_start` altogether from Iteration API and advise developers to use a custom callback based on either `cursor_position` accessor, or the business logic.

This is kind of related to https://github.com/Shopify/job-iteration/issues/19.

cc @GustavoCaso

**Comment by pedropb** [comment]
Hey 👋 

We ran into a similar problem in Russbot and had a discussion in our team about this.
Personally I like the `on_start` and `on_complete` callbacks. Given the job interruption/resume lifecycle is managed by the library, I think there's value in injecting behaviour at specific points in the lifecycle (`on_start`, `on_resume`, `on_interrupt`, `on_complete`).

If I may, an approach I'd like to suggest is:
1. the `cursor_position` job metadata is owned by JobIteration and should not be manipulated by user code
2. `run_callback(:start) if cursor_position.nil?`
3. if the user code wants to enqueue a job at a specified position, they should implement something like:
```ruby
# Job.perform_later(initial_cursor_position: desired_position)
on_start -> { job.cursor_position = job.initial_cursor_position }  
# Or alternatively, JobIteration API could consume a well defined `initial_cursor_position` job argument.
```

There's one edge case with CustomEnumerators that do not yield cursors (i.e: the [redis enumerator example from the docs](https://github.com/Shopify/job-iteration/blob/master/guides/custom-enumerator.md)). Enumerators like this usually are unbounded and don't really distinguish between resume and retry semantics - they just consume from a queue, or run through an ever changing set of records.

If we still wanted to provide the callbacks for cases like this, we could do something like:
```ruby
run_callbacks(:start) if cursor_position.nil? && times_interrupted == 0
```

This is a breaking change ofc and would cause problems if existing user code relied on manipulating the cursor_position and the on_start callback at the same time.

------------------

I'm happy to create a PR, if you think it's a good idea. It would also close #19, I think.

**Comment by grcooper** [comment]
Some thoughts:

I agree that the `on_start` and `on_complete` callbacks are great tools, but they have some caveats that make them non-intuitive (as explained above).

If we could keep them that would be great for the simple case, developers may also end up running into problems with writing their own `on_start` callbacks in similar ways. However, this would make it opt-in for them if they do decide to do that.

For @pedropb 's suggestion:
Agreed the cursor_position should not be modified by user code. However, I think making the user modify the `cursor_position` in their code goes against that, so why is that the suggestion laid out? This means that users have to know the internals of job_iteration just to enqueue a job at a specific cursor. Also as you mentioned, this is a breaking change.

Two possible solutions I might see are:
1. Add a flag for the callbacks `on_start_ran` (or something along those lines) which holds boolean flags for these specific callbacks to make sure that they are executed. These can be stored in the job payload similar to `cursor_position` and now they don't rely on any external factors, only those set and controlled by the iteration api.

2. Fix concurrency. If concurrency is re-enqueing the job before it has actually "executes" then perhaps there is something we have mis-represented there. It might be better to fix it at the concurrency layer and either adjust the amount of executions that have occurred, or change where we push things back into the queue so executions don't increase.


1 seems viable, 2 I have not done enough digging to investigate yet, but I suspect those might be better approaches if we want to leave the `on_start` callback.

**Comment by pedropb** [comment]
> However, I think making the user modify the cursor_position in their code goes against that, so why is that the suggestion laid out?

I left that suggestion just to build up the argument for an `initial_cursor_position` argument as part of the API.
```ruby
backfill_job = ImportJob.perform_later(initial_cursor_position: 1234)  
# 1234 is captured by JobIteration and set to cursor_position before the first iteration of the job
```
